### PR TITLE
KAFKA-12648: Pt. 3 - addNamedTopology API

### PR DESCRIPTION
**Pull request by ableegoldman**
Pt. 1: [#10609](https://github.com/apache/kafka/pull/10609)
Pt. 2: [#10683](https://github.com/apache/kafka/pull/10683)
Pt. 3: [#10788](https://github.com/apache/kafka/pull/10788)

In Pt. 3 we implement the `addNamedTopology` API. This can be used to update the processing topology of a running Kafka Streams application without resetting the app, or even pausing/restarting the process. It's up to the user to ensure that this API is called on every instance of an application to ensure all clients are able to run the newly added NamedTopology. This should not be too much of a burden as it only requires that each client eventually be updated by the user -- under the covers, Streams will take care of keeping the internal state consistent while various clients wait to converge on the latest view of the full topology.

Internally, when a new NamedTopology is added a rebalance will be triggered to distribute the tasks that correspond to it. To minimize disruption and wasted work, the assignor just computes the desired eventual assignment of these new tasks to clients regardless of whether the target client has been issued the `addNamedTopology` request yet. If a client receives tasks for a NamedTopology it does not yet recognize, it stashes them away and continues to process its other topologies. Once it receives this new NamedTopology, those tasks will be created and begin processing without triggering a new rebalance. If the new NamedTopology does not match any unknown tasks it has received, then the client must trigger a fresh rebalance for this new NamedTopology.

### CODE REVIEW

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/NamedTopologyTest.java
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/NamedTopologyTest.java:112]: We changed the behavior to return an empty Optional rather than throw, as users may want to use this API to determine whether the given named topology is known or not

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java]: nit: extra space.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java]: Use the newly added `Utils.filterMap`?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTaskCreator.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTaskCreator.java]: Ditto here.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/KafkaStreamsNamedTopologyWrapper.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/KafkaStreamsNamedTopologyWrapper.java]: ".. since the the topology name {} cannot be recognized"

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java]: I feel a bit concerned about the "asymmetry" of this function: all other functions have the lock inside while this function is supposed to be called by a caller -- i.e. `handleTopologyUpdatesPhase`. It is quite vulnerable to bugs with additional edits.

I'm wondering if we can move this logic out of `handleTopologyUpdatesPhase` instead: i.e. we first update the named topology, and then based on the new version we can either wait or re-subscribe and trigger rebalance. WDYT?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java]: Also I'm wondering how that would be possible: if the thread/consumer has not updated their subscriptions, the leader should not assign any of the tasks of the newly added topics to it right?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: See my other comment: I'm wondering if we move `topologyMetadata.maybeWaitForNonEmptyTopology();` out of this function do we still need to grab the lock of topologyMetadata? Seems we can wrap the remaining getters into a single call. I.e.:

1. Update the topology version if possible (only need a getter to the topologyMetadata). Returns a boolean indicating if the topology has changed.
2. Check on the current topology, if it is empty first `handleEmptyTopology` and then `maybeWaitForNonEmptyTopology`.
3. Otherwise of 2), check whether 1) returned true, if yes call `taskManager.maybeCreateTasksFromNewTopologies` (btw I'm not sure if the topology has changed, i.e. the consumer subscription have not been updated, if this thread/consumer could ever get tasks of those unknown-yet topologies? I think the leader would not assign tasks to those who do not subscribe to yet); and also subscribe consumer and enforce rebalance.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java]: This seems only used for the `updatePhase` outside the class itself -- I think we should better just still keep it only used within a single function of the TopologyMetadata. See my other related comments.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: Similar to my other comment: I think we can just add a single function of `topologyMetadata` which returns is synchronized on the lock, instead of relying on the caller to grab lock? More specifically, the `topologyMetadata().hasNamedTopologies()` can be replaced with the `!currentNamedTopologies.isEmpty()` and as long as `namedTopologiesView()` is synchronized that should be sufficient?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java:63]: How about moving this into the TopologyVersion to be guarded within the same lock as well, instead of making itself a concurrent data structure?

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java:63]: Also we can optimize the condition on whether we should enforce a new rebalance per-thread in `StreamThread` as: `builders.keySet().equals(assignedNamedTopologies)`. The former is the set of topologies this instance knows, the latter is the set of topologies the leader knows, when they are the same there's no need to trigger rebalance.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java]: Could we print the `inputTopics` and `inputPatterns` as well in the message?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java:193]: Just to clarify: we either support N topologies where none of them have global topologies, or we support just one topology with only global topology? What's the rationale for supporting the second scenario?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientState.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientState.java]: `task {}`.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: The propagated `assignedNamedTopologies` is not used yet anywhere in the KS code. Could you elaborate a bit how it would be used in the future?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/KafkaStreamsNamedTopologyWrapper.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/KafkaStreamsNamedTopologyWrapper.java]: Also I checked on the ksql side and found that we first call `metadata.remove(queryId.toString());` and then this function, in this case would `getTopologyByName` always return nothing? cc @wcarlson5

File: streams/src/test/java/org/apache/kafka/streams/integration/NamedTopologyIntegrationTest.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/NamedTopologyIntegrationTest.java:76]: This is meta question: do we have coverage on scenarios where the leader/member's bookkept named-topologies set are different? I.e. 1) the leader would not try to create any tasks that it's own topology-metadata is not aware of even if other subscriptions contain more topics, 2) vice verse, the other members would not try to create tasks for assignment that its topology metadata does not recognize, while later when they get added the tasks gets created then?

**Review by guozhangwang — COMMENTED** [review]
Made a first pass on the PR.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java]: I think we discussed this already, but in case anyone else is wondering: the leader will always assign tasks based on its view of the current named topologies and topics, ie it does not check on individual subscriptions since the group is assumed to be eventually consistent in this regard. (Note this is actually no different than today; even if each instance of an app has a different input topic in their topology, they will all wind up receiving tasks for whichever topic the leader happened to have.)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/KafkaStreamsNamedTopologyWrapper.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/KafkaStreamsNamedTopologyWrapper.java]: Yes, that's how it should be - you need to remove the named topology before cleaning up its state. Just like you can't call `KafkaStreams#cleanUp` while the application is running.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java]: You mean print them separately? Sure

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java:193]: The second scenario is called "Kafka Streams" 😉  

(Users who aren't using named topologies should be able to use global stores as they do today)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: Removed until it's needed, if it is

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java]: This part of the code has been through a lot of refactoring and after the latest cleanup, I agree we should be able to avoid exposing it at all

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: The lock was less for thread-safety purposes and more about atomicity w.r.t updating some bookkeeping/state. But I guess I removed a lot of that bookkeeping recently, so I'll do another pass and clean up this handling

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java]: Yeah I left this comment before syncing with @ableegoldman offline, so it's better to clarify for others reading this PR :)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java]: The diff between `assignedActiveTasks` and `activeTasksToCreate` are those tasks that are already owned and hence do not need to be assigned, am I reading that right (ditto for standbys)? Why these should be considered as remove revoked?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/KafkaStreamsNamedTopologyWrapper.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/KafkaStreamsNamedTopologyWrapper.java]: The topology was not included? :)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: Not sure I understand this logic.. could you explain a bit? Also there's no unfrozen function so once a task is frozen it would stick with that state forever?

**Review by guozhangwang — COMMENTED** [review]
We a pass on the new commits, but I'm a bit confused on some logic around unknown task removal and task freezing.. maybe we can chat again for me to get your thoughts?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java]: Whoops, that's what happens when you push changes late at night 🙃 I'll fix it up...hopefully will make more sense then

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: This is another stupid bug...it should be freezing them if they _don't _ appear in the current topology. Tasks from removed named topologies are frozen to prevent them from processing any more records, so yes, it is meant to be permanent.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java]: Agreed, let me clean this up

**Comment by ableegoldman** [comment]
@guozhangwang I think I've addressed all your feedback and significantly cleaned up the streamthread event loop + topology locking, let me know if there's anything else

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: I refactored this part of the code quite a bit, let me know if you have any remaining concerns with the new logic

File: streams/src/test/java/org/apache/kafka/streams/integration/NamedTopologyIntegrationTest.java
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/integration/NamedTopologyIntegrationTest.java:76]: I'm still filling out the integration test suite, especially the multi-node testing, but I'll make sure this scenario has coverage. This will probably have to be in the followup Pt. 4 which expands `add/removeNamedTopology` to return a `Future`, since being able to block on this helps a lot with the testing.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java:63]: Well I removed the whole `assignedNamedTopologies` thing from this PR, but yes, that was one of the original intentions of adding it to the AssignmentInfo. I'm leaning towards keeping things simple for now and avoid adding on to the AssignmentInfo where possible. We can revisit this if the extra rebalances become a problem.

It could be sufficient to just use the set of actually-assigned topologies as a proxy. It's not perfect since it could be that the leader was aware of a topology and simply happened to not assign any of its tasks to this client, but I'd bet this gets rid of most of the unnecessary rebalances.

That said, I'm saving this optimization for a followup Pt. 4 PR. I'll revisit the best way to handle the `builders` map in that as well

**Comment by ableegoldman** [comment]
Java 8 & 11 tests passed, Java 16 build failure was unrelated: `Execution failed for task ':storage:unitTest'.`

**Review by wcarlson5 — APPROVED** [review]
@ableegoldman I still this this looks good!

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java:63]: SGTM.

What about the other comment, i.e. moving the `Map<String, InternalTopologyBuilder> builders` into the `TopologyVersion` itself? Besides the constructors, the only modifiers to `builders` seem to be `register/deregister`, in which we would always try to `getAndIncrement` version. So what about consolidating the modification of builders along with version bump, and hence we would not need to use a `ConcurrentNavigableMap`?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:883]: very nit: just add a TODO that we can improve this case to not always enforce rebalance when version bumped, in case we forgot in future PRs?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: How about moving this ahead of `pollPhase()`? We are likely to be kicked out of the group while blocked waiting here, so it's better to be aware of that and re-join the group immediately, rather than doing the restore/etc still which may be all wasted work.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: Just another thought: if this is just for the very short temporary phase between when the topology is removed to when the rebalance is finally triggered to remove the tasks (which should usually be in the next poll), could we just call `task.suspend` instead of adding the new `freeze` logic? When we finally close the tasks we can still transit from suspended to closed?

**Review by guozhangwang — COMMENTED** [review]
Made another pass.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: Well long story short I was trying to avoid mucking around in the task state management (which has historically been the source of many critical bugs) -- also we want to remove the `suspended` state soon anyways. But actually it seems simpler to just `close()` the tasks here and now altogether

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java:63]: > I'll revisit the best way to handle the builders map in that as well

I was referring to moving the `builders` map to the `TopologyVersion` in the above, ie I want to save that for Pt. 4  if that's ok

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Ack (although note that there's no wasted work on the restore phase since there's by definition nothing for the thread to do yet as it won't have been assigned any new tasks until it polls again). 

I don't think it really matters much where we put this for that reason, except for the case in which we start up with no topology -- then it's a waste to join the group in the first place, so we may as well wait until we receive something to work on. So yes, I'll move it back ahead of poll

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:778]: No actual changes here, just pulled the cleanup of tasks out into a separate new `#closeAndCleanUpTasks` method so we can call that on tasks from removed topologies

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java]: Making this debug since `warn` seems too intense, and I'm not sure it's even worthy of `info` -- also, with named topologies you would expect to see this almost every time a topology is removed since the thread will try to close those tasks as soon as it notices the topology's removal

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java:63]: > I was referring to moving the `builders` map to the `TopologyVersion` in the above, ie I want to save that for Pt. 4 if that's ok

Yup, totally.

File: streams/src/test/java/org/apache/kafka/streams/integration/NamedTopologyIntegrationTest.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/integration/NamedTopologyIntegrationTest.java:76]: SG.

**Comment by guozhangwang** [comment]
LGTM! Please feel free to merge after green builds.

**Comment by ableegoldman** [comment]
Just one unrelated test failure (reopened [KAFKA-13128](https://issues.apache.org/jira/browse/KAFKA-13128)): `StoreQueryIntegrationTest.shouldQueryStoresAfterAddingAndRemovingStreamThread`

**Comment by ableegoldman** [comment]
Merged to trunk -- thanks all for keeping up with the reviews so far 😄

**Pull request merged**
